pkg/alicloud/client/client_dns.go

```diff
@@ -40,16 +53,38 @@ func (f *clientFactory) NewDNSClient(region, accessKeyID, accessKeySecret string
 	}
 
 	return &dnsClient{
-		Client:            *client,
-		accessKeyID:       accessKeyID,
-		domainsCache:      f.domainsCache,
-		domainsCacheMutex: &f.domainsCacheMutex,
+		Client:                 *client,
+		accessKeyID:            accessKeyID,
+		domainsCache:           f.domainsCache,
+		domainsCacheMutex:      &f.domainsCacheMutex,
+		RateLimiter:            f.getRateLimiter(accessKeyID),
+		RateLimiterWaitTimeout: f.waitTimeout,
+		Logger:                 log.Log.WithName("ali-dnsclient"),
 	}, nil
 }
+func (f *clientFactory) getRateLimiter(accessKeyID string) *rate.Limiter {
+	// cache.Expiring Get and Set methods are concurrency-safe
+	// However, if f rate limiter is not present in the cache, it may happen that multiple rate limiters are created
+	// at the same time for the same access key id, and the desired QPS is exceeded, so use f mutex to guard against this
+
+	f.rateLimitersMutex.Lock()
+	defer f.rateLimitersMutex.Unlock()
+
+	// Get f rate limiter from the cache, or create f new one if not present
+	var rateLimiter *rate.Limiter
+	if v, ok := f.rateLimiters.Get(accessKeyID); ok {
+		rateLimiter = v.(*rate.Limiter)
+	} else {
+		rateLimiter = rate.NewLimiter(f.limit, f.burst)
+	}
+	// Set should be called on every Get with cache.Expiring to refresh the TTL
+	f.rateLimiters.Set(accessKeyID, rateLimiter, rateLimiterCacheTTL)
```

Comment:
Not investigate the source code of `rateLimiter`. Does `rateLimiters.Get` refresh the key? If so,  maybe `	f.rateLimiters.Set(accessKeyID, rateLimiter, rateLimiterCacheTTL)` can be placed in the else block.

Reply:
No, Get doesn't refresh the key but just return value of unexpired key